### PR TITLE
Async Iteration and down-level Generators

### LINKED ISSUE #1564: ES6 Generator support
This can be broken down to 3 areas:
1. basic (parser) support, even if we lose typechecking locally
2. Typechecking support
3. Transpiling like facebook/regenerator or Traceur, for targeting ES3/ES5.

Any timelines on this?

### LINKED ISSUE #4031: iterators with ES5 target
This is related to https://github.com/Microsoft/TypeScript/issues/3164 which implies that TypeScript will NEVER support using generic iterators when the target is ES5. I understand that supporting this requires design tradeoffs that the team wants to avoid. However, if this is not done, it's difficult to see how TypeScript can be considered a serious alternative to Babel and Traceur for ES6 support. It will be a long time (2 years?) until most web apps will have as their minimum browser requirement only browsers that support ES6 iterators. I really think the direction presented in issue 3164 needs to be reconsidered.

### LINKED ISSUE #11209: for-of with map.keys() gives TS2495
**TypeScript Version:**  2.0.3

**Code**

``` ts
let map = new Map<string, string>();
map.set('hello', 'world');

for (let key of map.keys()) {
    console.log(key);
}
```

**Expected behavior:**
It should not throw.

**Actual behavior:** 
It throws `error TS2495: Type 'IterableIterator<string>' is not an array type or a string type.`

**tsconfig.json:**

``` json
{
  "compilerOptions": {
    "experimentalDecorators": true,
    "lib": [
      "dom",
      "es7",
      "scripthost"
    ],
    "moduleResolution": "node",
    "noImplicitAny": false,
    "preserveConstEnums": true,
    "pretty": true,
    "removeComments": false,
    "sourceMap": true,
    "target": "ES5"
  }
}
```

cc: @Mischi

### LINKED ISSUE #11326: Asynchronous Iterations
The [asynchronous iterations](https://github.com/tc39/proposal-async-iteration) proposal is now in stage 3, so it should be safe to support that in TypeScript now. Currently, a function may be a generator or an async function, but it cannot be both. The asynchronous iteration proposal removes this restriction. Here's an example from their readme:

``` javascript
async function* readLines(path) {
  let file = await fileOpen(path);

  try {
    while (!file.EOF) {
      yield await file.readLine();
    }
  } finally {
    await file.close();
  }
}
```
## Types

Generator functions declare several interfaces:

``` typescript
interface IteratorResult<T> {
    done: boolean;
    value: T;
}
interface Iterator<T> {
    next(value?: any): IteratorResult<T>;
    return?(value?: any): IteratorResult<T>;
    throw?(e?: any): IteratorResult<T>;
}
interface Iterable<T> {
    [Symbol.iterator](): Iterator<T>;
}
interface IterableIterator<T> extends Iterator<T> {
    [Symbol.iterator](): IterableIterator<T>;
}
```

Async generators return promises instead of their `IteratorResult`, and use a new symbol, `Symbol.asyncIterator`:

``` typescript
interface IteratorResult<T> {
    done: boolean;
    value: T;
}
interface AsyncIterator<T> {
    next(value?: any): Promise<IteratorResult<T>>;
    return?(value?: any): Promise<IteratorResult<T>>;
    throw?(e?: any): Promise<IteratorResult<T>>;
}
interface AsyncIterable<T> {
    [Symbol.asyncIterator](): AsyncIterator<T>;
}
interface AsyncIterableIterator<T> extends AsyncIterator<T> {
    [Symbol.asyncIterator](): AsyncIterableIterator<T>;
}
```
## Emit

I guess that this would require another helper function to downlevel to ES2015, similar to the one used for async functions. Yield and await expression would both be replaced by yield expression, so they need to be decorated with some tag, like `yield ['await', promise]` for `await promise`.
## for-await-of

The proposal also introduces a new loop. An example from their readme:

``` javascript
for await (const line of readLines(filePath)) {
  console.log(line);
}
```

This iterates over the `[Symbol.asyncIterator]()` object and awaits the value of `.next()`, similar to a for-of loop. This loop can only be used in async functions.

### LINKED ISSUE #11970: Generator with Array Spread
`Version 2.1.0-dev.20161031`

```
const g = function*(){
    yield 1;
}

console.log([...g()]);
```

Compiles to the following which throws an error (treating the generator as an array, not an iterable):
```
var g = function* () {
    yield 1;
};
console.log(g().slice());
//# sourceMappingURL=build.js.map
```

Possible solution using `Array.from`:
```
var g = function* () {
    yield 1;
};
console.log(Array.from(g()).slice());
//# sourceMappingURL=build.js.map
```

### PR DESCRIPTION
**Pull request by rbuckton**
This PR adds full support for Async Generators and Async Iteration (`for..await..of`) as per the current [TC39 Proposal](https://github.com/tc39/proposal-async-iteration) for all script targets. It also contains full support for Generators in ES5/3, including `for..of`, spread, and iterator destructuring.

# Iterators

ES6 introduced `Iterator`, which is an object that exposes three methods, `next`, `return`, and `throw`, as per the following interface:
```ts
interface Iterator<T> {
  next(value?: any): IteratorResult<T>;
  return?(value?: any): IteratorResult<T>;
  throw?(e?: any): IteratorResult<T>;
}
```

This kind of iterator is useful for iterating over synchronously available values, such as the elements of an Array or the keys of a Map. An object that supports iteration is said to be "iterable" if it has a `Symbol.iterator` method that returns an `Iterator` object.

# Async Iterators

The Async Iteration proposal introduces an `AsyncIterator`, which is similar to `Iterator`. The difference lies in the fact that the `next`, `return`, and `throw` methods of an `AsyncIterator` return a `Promise` for the iteration result, rather than the result itself. This allows the caller to enlist in an asynchronous notification for the time at which the `AsyncIterator` has advanced to the point of yielding a value. An `AsyncIterator` has the following shape:

```ts
interface AsyncIterator<T> {
  next(value?: any): Promise<IteratorResult<T>>;
  return?(value?: any): Promise<IteratorResult<T>>;
  throw?(e?: any): Promise<IteratorResult<T>>;
}
```

An object that supports async iteration is said to be "iterable" if it has a `Symbol.asyncIterator` method that returns an `AsyncIterator` object.

# Generators

ES6 also introduced "Generators", which are functions that can be used to yield partial computation results via the `Iterator` interface and the `yield` keyword. Generators can also internally delegate calls to another iterable through `yield *`. For example:

```ts
function* f() {
  yield 1;
  yield* [2, 3];
}
```

# Async Generators

The Async Iteration proposal introduces "Async Generators", which are async functions that also can be used to yield partial computation results. Async Generators can also delegate calls via `yield*` to either an iterable or async iterable:

```ts
async function* f(p) {
  yield 1;
  await p;
  yield* [2, 3];
  yield* (async function *() { 
    await p; 
    yield 4; 
  })();
}
```

As with Generators, Async Generators can only be function declarations, function expressions, or methods of classes or object literals. Arrow functions cannot be Async Generators. Async Generators require a valid, global `Promise` implementation (either native or an ES6-compatible polyfill), in addition to a valid `Symbol.asyncIterator` reference (either a native symbol or a shim).

# The `for-await-of` Statement

Finally, ES6 introduced the `for..of` statement as a means of iterating over an iterable. Similarly, the Async Iteration proposal introduces the `for..await..of` statement to iterate over an async iterable:

```ts
async function* g() { ... }
async function f() {
  for await (const x of g()) {
  }
}
```

The `for..await..of` statement is only legal within an Async Function or Async Generator.

# Generators and Iteration for ES5/3

In addition, the following changes are included to bring full support for Generators to ES5/ES3:

* Added a `-downlevelIteration` compiler option, used when targeting ES5 or ES3. 
  * When set, the compiler uses new type check and emit behavior that attempts to call a `[Symbol.iterator]()` method on the iterated object if it is found, and creates a synthetic array iterator over the object if it is not. This requires a native `Symbol.iterator` or `Symbol.iterator` shim at runtime for any non-array values.
  * When not set, array destructuring, spread, and `for..of` are only supported for arrays as is the current behavior in TypeScript 2.1 and earlier.
* Array Destructuring now uses `Symbol.iterator` in ES5/3 if available when using `-downlevelIteration`, but can be used on an Array even if it does not define `Symbol.iterator` at run time or design time.
* Spread elements in Array, Call, and New expressions use `Symbol.iterator` in ES5/3 if available when using `-downlevelIteration`, but can be used on Array even if it does not define `Symbol.iterator` at run time or design time.
* `for..of` statements support `Symbol.iterator` in ES5/3 if available when using `-downlevelIteration`, but can be used on an Array even if it does not define `Symbol.iterator` at run time or design time.
* Async Functions **do not** require a `Symbol.iterator` polyfill at run time to operate.

# Related issues
* Fixes #11970 - Generator with Array Spread
* Fixes #11326 - Asynchronous Iterations
* Fixes #11209 - for-of with map.keys() gives TS2495
* Fixes #4031 - iterators with ES5 target
* Fixes #1564 - ES6 Generator support

### CODE REVIEW

**Comment by mischkl** [comment]
The amount of stuff this solves is amazing! Really really awesome!

**Comment by screendriver** [comment]
Awesome!! **Big** thank you!

**Comment by glen-84** [comment]
What is the milestone for this? Will it make it into 2.1?

**Comment by rbuckton** [comment]
@glen-84 This will likely be part of a future milestone, but will not be part of 2.1.

**Comment by benjamingr** [comment]
Given the meeting notes on low level generators - can anyone let us know what's the status of async iterators?

**Comment by benjamingr** [comment]
Also, maybe add it to [the roadmap](https://github.com/Microsoft/TypeScript/wiki/Roadmap) ?

**Comment by isiahmeadows** [comment]
What's the status of this PR?

**Comment by rbuckton** [comment]
@isiahmeadows Still in progress. I need to move the target for Async Generators to `esnext` as it does not look like it will be in ES2017.

**Comment by benjamingr** [comment]
@rbuckton correct, async generators are stage 3, they'd need to be stage 4 to be included - which means implementation in browsers and testing it which didn't happen yet. The Chrome PR is about to start review around Christmas and is not expected to land in the next two months anyway. ES2018 is a reasonable deadline though.

**Comment by benjamingr** [comment]
Note that we have created test262 tests over the weekend for this: https://github.com/benjamingr/test262/tree/AsyncGeneratorExpressions 

You might want to take a look.

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -175,23 +175,23 @@
         "category": "Error",
         "code": 1057
     },
-    "Operand for 'await' does not have a valid callable 'then' member.": {
+    "Type used as operand to 'await' or the return type of an async function must not contain a callable 'then' member if it is not a promise.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: This double negative at the end is hard to read. Would it be accurate to reword as " ... may only contain a callable 'then' member if it is a promise." ?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -203,26 +198,20 @@ namespace ts {
...
+        let deferredGlobalPromiseType: GenericType;
+        let deferredGlobalPromiseConstructorSymbol: Symbol;
+        let deferredGlobalPromiseConstructorLikeType: ObjectType;
+        let deferredGlobalIterableType: GenericType;
+        let deferredGlobalIteratorType: GenericType;
+        let deferredGlobalIterableIteratorType: GenericType;
+        let deferredGlobalAsyncIterableType: GenericType;
+        let deferredGlobalAsyncIteratorType: GenericType;
+        let deferredGlobalAsyncIterableIteratorType: GenericType;
+        let defferedGlobalTemplateStringsArrayType: ObjectType;
```
Comment (sandersn) [inline src/compiler/checker.ts]: typo: deffered -> deferred

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -5650,20 +5647,75 @@ namespace ts {
...
+            return getGlobalSymbol(name, SymbolFlags.Type, unchecked ? undefined : Diagnostics.Cannot_find_global_type_0);
         }
 
         function getGlobalSymbol(name: string, meaning: SymbolFlags, diagnostic: DiagnosticMessage): Symbol {
             return resolveName(undefined, name, meaning, diagnostic, name);
         }
 
-        function getGlobalType(name: string, arity = 0): ObjectType {
-            return getTypeOfGlobalSymbol(getGlobalTypeSymbol(name), arity);
+        function getGlobalType(name: string, arity?: 0, unchecked?: boolean): ObjectType;
```
Comment (sandersn) [inline src/compiler/checker.ts]: I think `reportErrors` would be a better name for `unchecked`. Or at least `checked`, like the majority of uses in this PR.

Comment (sandersn) [inline src/compiler/checker.ts]: Also, can you explain the introduction of the checked/unchecked parameter? Why do some callers want an error and others not? Would it work for the checked callers to issue the error themselves?

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -2073,9 +2074,7 @@ namespace ts {
             for (let i = 0; i < lines.length; i++) {
                 const line = indentation ? lines[i].slice(indentation) : lines[i];
                 if (line.length) {
-                    if (i > 0) {
-                        writeLine();
-                    }
+                    writeLine();
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: Looks like a merge issue, but this doesn't result in multiple lines. `writeLine()` doesn't actually write a line, it just sets a flag to write a line before the next `write()`, so multiple lines are always reduced to a single line.

File: src/compiler/checker.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: I am trying to move away from the `memoize` approach to better report errors in vscode. In some cases we don't want to report an error if we can't find the type, such as with down-level async functions.

File: src/compiler/binder.ts [hunk 5]
```diff
@@ -3138,9 +3137,18 @@ namespace ts {
                 transformFlags |= TransformFlags.AssertES2015;
                 break;
 
+            case SyntaxKind.ForOfStatement:
+                // This node is either ES2015 syntax or ES2017 syntax (if it is a for-await-of).
+                if ((<ForOfStatement>node).awaitModifier) {
+                    transformFlags |= TransformFlags.AssertESNext;
```
Comment (rbuckton) [inline src/compiler/binder.ts]: The way that the esnext transformer works now is to always descend through the tree if *some* node is marked `ESNext`. The `SpreadAssignment` case handles that further down, so we can reduce false positives in `for..of` triggering the esnext transform when it isn't needed.

File: Gulpfile.ts [hunk 6]
```diff
@@ -136,6 +136,14 @@ const es2017LibrarySourceMap = es2017LibrarySource.map(function(source) {
     return { target: "lib." + source, sources: ["header.d.ts", source] };
 });
 
+const esnextLibrarySource = [
+    "esnext.asynciterable.d.ts"
+];
+
+const esnextLibrarySourceMap = esnextLibrarySource.map(function (source) {
```
Comment (rbuckton) [inline Gulpfile.ts]: Consistency with the surrounding code.

File: src/compiler/checker.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: I used `unchecked` here (or I could use a `doNotReportErrors`, though that's a bit wordy) to preserve existing call sites that do not require an `arity` check.

File: src/compiler/comments.ts [hunk 7]
```diff
@@ -262,7 +262,8 @@ namespace ts {
 
         function forEachLeadingCommentToEmit(pos: number, cb: (commentPos: number, commentEnd: number, kind: SyntaxKind, hasTrailingNewLine: boolean, rangePos: number) => void) {
             // Emit the leading comments only if the container's pos doesn't match because the container should take care of emitting these comments
-            if (containerPos === -1 || pos !== containerPos) {
+            if ((containerPos === -1 || pos !== containerPos) /* && !leadingCommentPositions[pos] */) {
+                // leadingCommentPositions[pos] = true;
```
Comment (sandersn) [inline src/compiler/comments.ts]: This commented out code should also be deleted, right?

**Comment by sandersn** [comment]
>             // TODO(rbuckton): Verify whether we need to call getApparentType. See checkNonThenableType in master

turns out getTypeOfPropertyOfType calls getApparentType for you. So the answer is no.

---
Refers to: src/compiler/checker.ts:16496 in f9999e9. [](commit_id = f9999e9738bc038e75c5cfb5e6955f05025a0ad0, deletion_comment = False)

**Comment by sandersn** [comment]
>                 return typeAsAwaitable.awaitedTypeOfType = getUnionType(types, /*subtypeReduction*/ true);

could you replace the loop with mapType? You'd have to add a subTypeReduction parameter to mapType and figure out how to replicate the undefined filtering of `append` some other way.

---
Refers to: src/compiler/checker.ts:16595 in f9999e9. [](commit_id = f9999e9738bc038e75c5cfb5e6955f05025a0ad0, deletion_comment = False)

**Comment by sandersn** [comment]
>             const widenedType = getWidenedType(type);

is `getWidenedType` needed here? All it really does it change undefined/null/unknown to any, so it shouldn't change the answer of `isThenableType`.

---
Refers to: src/compiler/checker.ts:16667 in f9999e9. [](commit_id = f9999e9738bc038e75c5cfb5e6955f05025a0ad0, deletion_comment = False)

**Comment by sandersn** [comment]
>                     // promise. If so, we returnundefined as we cannot guess the shape. If this

typo: add space

---
Refers to: src/compiler/checker.ts:16603 in f9999e9. [](commit_id = f9999e9738bc038e75c5cfb5e6955f05025a0ad0, deletion_comment = False)

**Comment by rbuckton** [comment]
>                 return typeAsAwaitable.awaitedTypeOfType = getUnionType(types, /*subtypeReduction*/ true);

That sounds more complicated than what I'm doing now.

---
In reply to: [273871123](https://github.com/Microsoft/TypeScript/pull/12346#issuecomment-273871123) [](ancestors = 273871123)

---
Refers to: src/compiler/checker.ts:16595 in f9999e9. [](commit_id = f9999e9738bc038e75c5cfb5e6955f05025a0ad0, deletion_comment = False)

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -16374,30 +16477,37 @@ namespace ts {
...
+            //  {
+            //      then( // thenFunction
+            //      ): any;
+            //  }
+            //
 
-                return unknownType;
+            const thenFunction = getTypeOfPropertyOfType(type, "then");
+            const thenSignatures = thenFunction ? getSignaturesOfType(thenFunction, SignatureKind.Call) : emptyArray;
+            if (thenSignatures.length > 0) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: Why not just `return thenSignatures.length > 0` ?

File: src/compiler/transformers/es2017.ts [hunk 9]
```diff
@@ -317,69 +312,90 @@ namespace ts {
             }
         }
 
+        /**
+         * Hook for node emit.
+         *
+         * @param node The node to emit.
```
Comment (sandersn) [inline src/compiler/transformers/es2017.ts]: these `@param`s (and the ones from the next function) are out of date.

File: src/compiler/transformers/es2015.ts [hunk 10]
```diff
@@ -2362,18 +2379,105 @@ namespace ts {
...
+            const errorRecord = createUniqueName("e");
+            const catchVariable = getGeneratedNameForNode(errorRecord);
+            const returnMethod = createTempVariable(/*recordTempVariable*/ undefined);
+            const values = createValuesHelper(context, expression, node.expression);
+            const next = createCall(createPropertyAccess(iterator, "next" ), /*typeArguments*/ undefined, []);
+
+            hoistVariableDeclaration(errorRecord);
+            hoistVariableDeclaration(returnMethod);
+
+            const forStatement = setEmitFlags(
```
Comment (sandersn) [inline src/compiler/transformers/es2015.ts]: why is this code almost identical to the same transformation in esnext for `for await .. of`? I thought that the esnext transform could just let the es2015 transform take care of it.

File: src/compiler/checker.ts [hunk 11]
```diff
@@ -203,26 +198,20 @@ namespace ts {
...
         // This allows users to just specify library files they want to used through --lib
         // and they will not get an error from not having unrelated library files
-        let getGlobalTemplateStringsArrayType: () => ObjectType;
-
-        let getGlobalESSymbolType: () => ObjectType;
-        let getGlobalIterableType: () => GenericType;
-        let getGlobalIteratorType: () => GenericType;
-        let getGlobalIterableIteratorType: () => GenericType;
-
-        let getGlobalClassDecoratorType: () => ObjectType;
```
Comment (rbuckton) [inline src/compiler/checker.ts]: It was unused in the checker.

File: src/compiler/checker.ts [hunk 12]
```diff
@@ -17698,41 +17989,43 @@ namespace ts {
...
+                return undefined;
             }
 
-            return typeAsIterable.iterableElementType;
+            return typeAsIterable.iteratedTypeOfIterable = getIteratedTypeOfIterator(getUnionType(map(iteratorMethodSignatures, getReturnTypeOfSignature), /*subtypeReduction*/ true), errorNode);
         }
 
         /**
-         * This function has very similar logic as getElementTypeOfIterable, except that it operates on
+         * This function has very similar logic as getIteratedTypeOfIterable, except that it operates on
```
Comment (rbuckton) [inline src/compiler/checker.ts]: The `getIteratedTypeOf*` functions aren't repetitive, just similarly structured. Each function has a different responsibility. Having a single `getIteratedType` function with various parameters to control its behavior would conflate all of those responsibilities and result in a very large function.

* `getIteratedTypeOfIterator` - gets the iterated type from an `Iterator<T>`-like type, with a fast path for type references to `Iterator<T>`.
* `getIteratedTypeOfIterable` - gets the iterated type from an `Iterable<T>`-like type, with a fast path for type references to `Iterable<T>` and `IterableIterator<T>`.  If the type has a `Symbol.iterator` method, it relies on `getIteratedTypeOfIterator`.
* `getIteratedTypeOfIterableIterator` - gets the iterated type from an `Iterable<T>`-like type or an `Iterator<T>`-like type. The order it resolves the type is important as this is used to verify the return type of a generator function.

* `getIteratdTypeOfAsyncIterator` - While similar to `getIteratedTypeOfIterator`, this function specifically targets `AsyncIterator<T>` and gets the promised type of the `next()` method. This function uses a different cache location for the resulting type, so any attempt to merge these functions would likely result in a significant amount of bifurcation.
* `getIteratedTypeOfAsyncIterable` - While similar to `getIteratedTypeOfIterable`, this function specifically targets `AsyncIterable<T>`. As with `getIteratedTypeOfAsyncIterator`, merging this function with its non-async counterpart would result in significant bifurcation due to how conflicts are cached.

* `getIteratedTypeOfIterableOrElementTypeOfArray` - used to get the iterated type in a for-of statement for `--target es5`.
* `getIteratedTypeOfIterableOrElementTypeOfArrayOrString` - used as part of checking the iterated type of the right-hand-side of a for-of statement where `string` may be allowed.
* `getIteratedTypeOrElementType` - conditionally calls `getIteratedTypeOfIterable`, `getIteratedTypeOfIterableOrElementTypeOfArray` or `getIteratedTypeOfIterableOrElementTypeOfArrayOrString` based on passed in requirements.

While this may seem like a "thicket", each function has a specific responsibility and I think is generally more maintainable than a monolithic function that tries to serve far too many needs.

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -17822,8 +18113,14 @@ namespace ts {
          *   1. Some constituent is neither a string nor an array.
          *   2. Some constituent is a string and target is less than ES5 (because in ES3 string is not indexable).
          */
-        function checkElementTypeOfArrayOrString(arrayOrStringType: Type, errorNode: Node): Type {
-            Debug.assert(languageVersion < ScriptTarget.ES2015);
+        function getIteratedTypeOfIterableOrElementTypeOfArrayOrString(arrayOrStringType: Type, errorNode: Node, checkAssignability: boolean): Type {
+            const iteratedType = compilerOptions.downlevelIteration && getIteratedTypeOfIterable(arrayOrStringType, /*errorNode*/ undefined);
+            if (iteratedType) {
+                if (checkAssignability && errorNode) {
+                    checkTypeAssignableTo(arrayOrStringType, createIterableType(iteratedType), errorNode);
```
Comment (rbuckton) [inline src/compiler/checker.ts]: This is a check for `--downlevelIteration`. We perform the check for the same reason as mentioned in `checkIteratedTypeOfIterableOrAsyncIterable`:
> even though we have extracted the iteratedType, we will have to validate that the type passed in is actually an Iterable.

This catches a case where our `getIteratedTypeOfIterable` resolution finds an iterated type, but the array or string type is not actually assignable to `Iterable<T>`, reporting an assignability error.

File: src/compiler/checker.ts [hunk 14]
```diff
@@ -17880,6 +18186,26 @@ namespace ts {
             return arrayElementType;
         }
 
+        function getIteratedTypeOfIterableOrElementTypeOfArray(inputType: Type, errorNode: Node, checkAssignability: boolean): Type {
+            const iteratedType = compilerOptions.downlevelIteration && getIteratedTypeOfIterable(inputType, /*errorNode*/ undefined);
```
Comment (rbuckton) [inline src/compiler/checker.ts]: `getIteratedTypeOfIterableOrElementTypeOfArrayOrString` has different logic than `getIteratedTypeOfIterableOrElementTypeOfArray`, we don't want to delegate.

**Comment by klihelp** [comment]
thanks!

File: src/compiler/transformers/es2015.ts (hunk 10 above)
Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: If you use `--target es2016`, then the es2015 transform will never be run.

File: src/compiler/transformers/es2017.ts [hunk 15]
```diff
@@ -450,15 +430,26 @@ namespace ts {
...
-                && resolver.getNodeCheckFlags(currentSuperContainer) & (NodeCheckFlags.AsyncMethodWithSuper | NodeCheckFlags.AsyncMethodWithSuperBinding);
-        }
     }
 
+    const awaiterHelper: EmitHelper = {
+        name: "typescript:awaiter",
+        scoped: false,
+        priority: 5,
+        text: `
+            var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
```
Comment (rbuckton) [inline src/compiler/transformers/es2017.ts]: We don't use a minified representation for these names currently.

File: tests/baselines/reference/types.forAwait.esnext.1.symbols [hunk 16]
```diff
@@ -0,0 +1,55 @@
+=== tests/cases/conformance/types/forAwait/types.forAwait.esnext.1.ts ===
```
Comment (rbuckton) [inline tests/baselines/reference/types.forAwait.esnext.1.symbols]: Because they have `// @noEmit: true` at the top of the file. I am not checking emit with these tests, just type resolution. The .js output would be extraneous. There should be enough coverage in `emitter.forAwait.esnext.ts`, `emitter.forAwait.es2015.ts`, and `emitter.forAwait.es5.ts` for that.

File: src/compiler/checker.ts (hunk 12 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: I went ahead and merged the functions.

**Comment by TonyPythoneer** [comment]
I am expecting this PR will merge into stable or RC version.

**Comment by TonyPythoneer** [comment]
This PR is planed on TypeScript 2.3 release. :sob:

**Comment by rbuckton** [comment]
@mhegazy, Is there anyone else you want to review this or should I merge?

File: Jakefile.js [hunk 17]
```diff
@@ -170,13 +170,21 @@ var es2016LibrarySourceMap = es2016LibrarySource.map(function (source) {
...
-    "es2017.string.d.ts",
+    "es2017.string.d.ts"
 ];
 
 var es2017LibrarySourceMap = es2017LibrarySource.map(function (source) {
     return { target: "lib." + source, sources: ["header.d.ts", source] };
 });
 
+var esnextLibrarySource = [
+    "esnext.asynciterable.d.ts"
```
Comment (mhegazy) [inline Jakefile.js]: we need to update VS setup to include this file as well.

**Comment by jocull** [comment]
Just curious, why is the `downlevelIteration` still undocumented under `tsc --help`? It seems to work, but there's no mention of it, and the errors leading up to needing it are unhelpful.

**Comment by mhegazy** [comment]
> Just curious, why is the downlevelIteration still undocumented under tsc --help?

it is, just use `tsc --help --all`
